content/docs/grammar-language.md

```diff
@@ -0,0 +1,371 @@
+
+#### Negated Token
+It is possible to negate all tokens using the operator `!`, in regular expression the operator `^` is used for negation.
+```
+terminal BETWEEN_HASHES:'#' (!'#')* '#';
+```
+In regular expression:
+```
+terminal BETWEEN_HASHES: /#[^#]*#/;
+```
+
+#### Rule Calls
+A terminal rule can implement other terminal rules. 
+```
+terminal DOUBLE returns number: INT '.' INT;
+```
+Note: it is easy to create conflict between terminal rules by using *terminal rule calls*. See [Data Type Rules](#data-type-rules) for further details.
+
+#### Alternatives
+It is possible to match several valid options. The terminal rule `STRING` can use alternatives to match a sequence of characters between double quotes `""` or single quotes `''`:
+```
+terminal STRING: /"[^"]*"|'[^']*'/;
+```
+
+#### Groups
+Tokens can be put in sequence specifying the order they have to appear
+```
+terminal FLIGHT_NUMBER: ('A'..'Z')('A'..'Z')('0'..'9')('0'..'9')('0'..'9')('0'...'9')?;
+```
+In this example, the token `FLIGHT_NUMBER` **must** start with two capital letters and followed by three of four digits.
+
+### Terminal Fragments
+Fragments allow for sub-definition of terminal rules to be extracted. They are not consumed by the lexer and have to be consumed by other terminal rules.
+```
+terminal fragment CAPITAL_LETTER: ('A'..'Z');
+terminal fragment SMALL_LETTER: ('a'..'z');
+terminal NAME: CAPITAL_LETTER SMALL_LETTER+;
+```
+In this example, the lexer will not transform a single capital or small letter into a valid token but will match a sequence of one capital letter followed by one or many small letters.
+
+### Hidden Terminal Rules
+The parser tries to match every character in the document to a terminal rules. It is therefore necessary to specify which characters or sequence of characters need to be ignored during parsing. Generally, you would want to ignore whitespaces and comments. This is achieved by adding the keyword `hidden` in front of the keyword `terminal` when defining a *terminal rule*. These *hidden terminal rules* are global and will be valid for all parser rules in the document.
+```
+hidden terminal WS: /\s+/;
+hidden terminal ML_COMMENT: /\/\*[\s\S]*?\*\//;
+hidden terminal SL_COMMENT: /\/\/[^\n\r]*/;
+```
+
+## Parser Rules
+While [terminal rules](#terminal-rules) indicate to the lexer what sequence of characters are valid tokens, *parser rules* indicate to the parser what sequence of tokens are valid. Parser rules lay the structure of objects to be created by the parser and result in the creation of the *abstract syntax tree* which represents the syntactic structure of your language. In Langium, parser rules are also responsible for defining the type of objects to be parsed.
+
+### Extended Backus-Naur Form Expressions
+Parser rules have access to a more restricted set of EBNF expressions compared to terminal rules. Parser and terminal rules share the following expressions:
+- Groups
+- Alternatives
+- Keywords
```

Comment:
Terminal rules are also able to use keywords. `terminal X: 'Y';` is a valid terminal.

Reply:
Right, but we don't call those keywords. They are merely character sequences (like groups): `'xyz'` is the same as `'x' 'y' 'z'`.